lib/provisioner.rb

```diff
@@ -196,6 +200,162 @@ def garbage_collect
+        credentialsid = doc.at_css "credentialsId"
+        credentialsid.content = credentials_id
+        jobXml = doc.to_html
+        job = ::JenkinsApi::Client::Job.new(client)
+        job_created = false
+        for counter in 0..20
+          begin
+            job.create_or_update(agent_name, jobXml)
+            job_created = true
+            break
+          rescue
+            STDERR.puts $!, $@ # Print exception
+            sleep 5
+            next
+          end
+        end
+        if job_created
+          STDOUT.puts "Registered Shared Slave Successfully"
+        else
+          raise SharedSlaveNotCreatedError, "Shared Slave not Created."
+        end
+        sleep @@registration_wait_time
+      end
+    end
+
+    def deleteJobs(vm_hashes)
+      STDOUT.puts "Deleting all jobs on Jenkins Operation Center."
+      jenkins_username = @configuration.jenkins_username
+      jenkins_password = @configuration.jenkins_password
+      jenkins = @configuration.jenkins
+      private_key_path = @configuration.private_key_path
+      credentials_id = @configuration.credentials_id
+      labels = @configuration.labels
+      client = ::JenkinsApi::Client.new(:username => jenkins_username,
+                                        :password => jenkins_password, :server_url => jenkins)
+      vm_hashes.each do |vm_hash|
+        counter = 0
+        agent_ip = vm_hash['TEMPLATE']['NIC']['IP']
+        agent_name = "agent - #{agent_ip}"
+        #First disable job.
+        jobXml = File.open("slave.xml")
+        doc = Nokogiri::XML(jobXml)
+        disabled  = doc.at_css "disabled"
+        disabled.content = true
+        jobXml = doc.to_html
+        job = ::JenkinsApi::Client::Job.new(client)
+        job_created = false
+        for counter in 0..20
+          begin
+            job.create_or_update(agent_name, jobXml)
+            job_created = true
+            break
+          rescue
+              STDERR.puts $!, $@ # Print exception
+              sleep 5
+              next
+          end
+        end
+        if job_created
+          STDOUT.puts "Disabled Job Successfully"
+        else
+          raise SharedSlaveNotCreatedError, "Shared Slave not Disabled."
+        end
+        sleep @@registration_wait_time
+        #delete job once disbaled.
+        begin
+          job.delete(agent_name) # delete job if exists
+        rescue
+          next
+        end
+        STDOUT.puts "Deleted Job Successfully"
+        sleep @@registration_wait_time
+      end
+    end
+
+    ##
+    # Garbage collection means asking Jenkins what nodes it currently has and then performing
+    # the cleanup on the OpenNebula side. Don't clean anything that is younger than 5 minutes.
+
+    def garbage_collect
+      jenkins = @configuration.jenkins
+      jenkins_username = @configuration.jenkins_username
+      jenkins_password = @configuration.jenkins_password
+      endpoint = (jenkins[-1] == 47 || jenkins[-1] == '/') ?
```

Comment:
Maybe just use end_with? instead of doing this character checking. 

(no need to change it - just a comment)